src/physics_swe.c

```diff
@@ -296,58 +296,51 @@ static PetscErrorCode RHSFunctionForInternalEdges(RDy rdy, Vec F, CourantNumberD
   PetscFunctionReturn(0);
 }
 
-// applies a reflecting boundary condition on the given boundary, computing
-// fluxes F for the solution vector components X
-static PetscErrorCode ApplyReflectingBC(RDy rdy, RDyBoundary *boundary, PetscReal tiny_h, CourantNumberDiagnostics *courant_num_diags, PetscReal *X,
-                                        PetscReal *F) {
+// Before computing BC fluxes, perform common precomputation irrespective of BC type that include:
+// (i) extracting h/hu/hv from the solution vector X, and
+// (ii) compute velocities (u/v) from momentum (hu/hv).
+static PetscErrorCode PerformPreComputationForBC(RDy rdy, RDyBoundary *boundary, PetscReal tiny_h, PetscInt N, PetscReal hl[N], PetscReal ul[N],
+                                                 PetscReal vl[N], PetscReal *X) {
   PetscFunctionBeginUser;
 
-  RDyCells *cells = &rdy->mesh.cells;
   RDyEdges *edges = &rdy->mesh.edges;
 
-  PetscInt num = boundary->num_edges;
-
-  PetscReal hl_vec_bnd[num], hul_vec_bnd[num], hvl_vec_bnd[num], ul_vec_bnd[num], vl_vec_bnd[num];
-  PetscReal hr_vec_bnd[num], ur_vec_bnd[num], vr_vec_bnd[num];
-  PetscReal sn_vec_bnd[num], cn_vec_bnd[num];
-  PetscReal flux_vec_bnd[num][3], amax_vec_bnd[num];
+  PetscReal hul[N], hvl[N];
 
   // Collect the h/hu/hv for left cells to compute u/v
   for (PetscInt e = 0; e < boundary->num_edges; ++e) {
     PetscInt iedge = boundary->edge_ids[e];
     PetscInt icell = edges->cell_ids[2 * iedge];
 
-    hl_vec_bnd[e]  = X[3 * icell + 0];
-    hul_vec_bnd[e] = X[3 * icell + 1];
-    hvl_vec_bnd[e] = X[3 * icell + 2];
+    hl[e]  = X[3 * icell + 0];
+    hul[e] = X[3 * icell + 1];
+    hvl[e] = X[3 * icell + 2];
   }
 
   // Compute u/v for left cells
-  PetscCall(GetVelocityFromMomentum(num, tiny_h, hl_vec_bnd, hul_vec_bnd, hvl_vec_bnd, ul_vec_bnd, vl_vec_bnd));
+  PetscCall(GetVelocityFromMomentum(N, tiny_h, hl, hul, hvl, ul, vl));
 
-  // Compute h/u/v for right cells
-  for (PetscInt e = 0; e < boundary->num_edges; ++e) {
-    PetscInt iedge = boundary->edge_ids[e];
-    PetscInt icell = edges->cell_ids[2 * iedge];
+  PetscFunctionReturn(0);
+}
 
-    cn_vec_bnd[e] = edges->cn[iedge];
-    sn_vec_bnd[e] = edges->sn[iedge];
+// After the right values (hr/ur/vr) have been computed based on the different type of BCs,
+// compute the fluxes and add contribution in the F vector.
+static PetscErrorCode ComputeBC(RDy rdy, RDyBoundary *boundary, PetscReal tiny_h, CourantNumberDiagnostics *courant_num_diags, PetscInt N,
```

Comment:
@jeff-cohere The input argument list is absurdly too long and ideas on how to make it more readable are welcome. Maybe we can make a struct that contains the three variables`h`, `u` and `v`.

Reply:
We could define a boundary condition struct that holds all this information, and initialize it with an appropriate function. For example, if we shoved all of the arrays of length `N` into a `SWEBCData` struct (for example):

```
typedef struct {
  // boundary associated with this data
  RDyBoundary *boundary;
  // number of boundary edges
  PetscInt N;
  // boundary edge data
  PetscReal *hl, *hr, *ul, *ur, *vl, *vr, *sn, *cn;
} BCDataSWE;

static PetscErrorCode BCDataSWECreate(RDyBoundary *boundary, BCDataSWE* bc_data) {
  PetscFunctionBegin;
  bc_data.boundary = boundary;
  PetscInt N = boundary->num_edges;
  bc_data.N = N;
  RDyAlloc(PetscReal, N, &boundary->hl);
  // etc
}

static PetscErrorCode BCDataSWEDestroy(BCDataSWE bc_data) {
  PetscFunctionBegin;
  RDyFree(bc_data.hl);
  // etc
}
```

Then `ComputeBC` might look like this:

```
static PetscErrorCode ComputeBC(RDy rdy, BCDataSWE* bc_data, PetscReal tiny_h, CourantNumberDiagnostics *courant_num_diags, PetscReal *X, PetscReal *F) {
  // etc
}
```

and `PerformPreComputationForBC` (or `PrecomputeBCData`, perhaps) would look similar.

Does this sound reasonable?

Reply:
Of course, it might be good to chat with Jed and Matt sometime soon about their libceed QFunctions and their vectorization, in case we need to do anything special for our GPU-friendly strategy.